app/controllers/concerns/import.rb

```diff
@@ -36,16 +36,25 @@ def import
     end
   end
   
-  def remove_codes
-    flag = offeror_codes.where(:user_id => nil)
-    if flag.count == 0
+  def remove_unclaimed_codes
+    if offeror_codes.where(user_id: nil).count.zero?
       redirect_to "/#{params[:controller]}/home",
         :flash => { :error => "There are no unclaimed codes" }
     else
-      contents = current_offeror.remove_unclaimed_codes(offeror_codes)
-      send_data contents, :filename => "unclaimed_codes.txt"
+      current_offeror.remove_unclaimed_codes(offeror_codes)
+      redirect_to "/#{params[:controller]}/home",
+        :flash => { :notice => "Codes were removed" }
     end
   end
+  
+  def download_unclaimed_codes
+    unclaimed_codes = current_offeror.download_unclaimed_codes(offeror_codes)
+    send_data(unclaimed_codes, filename: "unclaimed_codes.txt")
+  end
+  
+  def view_codes
+    @offeror_codes = offeror_codes.all
```

Comment:
Oleg, I've added before_action filters as a general solution to prevent invoking unpermitted actions by constructing requests.
In the case you've mentioned, during provider or vendor session, this request will show a page without codes (offeror_codes will return an empty collection). In other cases will occur redirection to the main page.